### PR TITLE
Se 875 tron inactive initiating action token trade level

### PR DESCRIPTION
**Pull request by TJ-Everett**

### CODE REVIEW

File: src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol [hunk 1]
```diff
@@ -41,23 +41,84 @@ contract ERC721NonTaggedRuleFacet is
...
+                    _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
+                }
+                /// tokenMinTxSize SELL Side 
+                if (lib.tokenMinTxSizeStorage().tokenMinTxSize[ActionTypes.SELL].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
+                /// tokenMaxDailyTrades SELL Side
+                if (lib.tokenMaxDailyTradesStorage().tokenMaxDailyTrades[ActionTypes.SELL].active) _checkTokenMaxDailyTradesRule(action, _tokenId);
+            } else { /// custodial buy 
+                if (lib.tokenMaxTradingVolumeStorage().tokenMaxTradingVolume[action]) {
+                    _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
+                }
```
Comment (VoR0220) [inline src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol:86]: can't shake the feeling that we can clean up the logic here. Same with ERC20 above. But it feels a little more surgical.

Comment (TJ-Everett) [reply src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol:86]: I am all ears on how - this one is different since it is more of an all or nothing check - since there is not active flag per action it has to be an else if - as far as I can tell.

Comment (oscarsernarosero) [reply src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol:86]: I agree. I think having a temporary action flag like the one RJ made for minHoldTime is a good idea:
```
bool P2P_trade = (action == ActionTypes.BUY && _from != _sender) || (action == ActionTypes.SELL && _sender != _to)
```
Then, you just check the rules as usual but you also ask about this action:
```
if(action == ActionTypes.BUY || P2P_trade)...
```
same for the SELL:
```
if(action == ActionTypes.SELL || P2P_trade)...
```

File: src/client/token/handler/diamond/ERC20NonTaggedRuleFacet.sol [hunk 2]
```diff
@@ -23,13 +23,67 @@ contract ERC20NonTaggedRuleFacet is AppAdministratorOrOwnerOnlyDiamondVersion, H
...
+                }
+                /// tokenMinTxSize BUY
+                if (lib.tokenMinTxSizeStorage().tokenMinTxSize[ActionTypes.SELL].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
+            } else { /// custodial buy 
+                if (lib.tokenMaxTradingVolumeStorage().tokenMaxTradingVolume[action]) {
+                    _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
+                }
+            }
+            /// tokenMinTxSize BUY
+            if (lib.tokenMinTxSizeStorage().tokenMinTxSize[action].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
```
Comment (VoR0220) [inline src/client/token/handler/diamond/ERC20NonTaggedRuleFacet.sol:64]: Ok, after staring at this in my editor (very hard to read here) have a suggestion:

``` 
if (_from != _sender){ /// non custodial buy 
                /// tokenMaxTradingVolume BUY 
                /// tokenMaxTradingVolume uses single rule id for all actions so check if Buy has rule id set ELSE check if sell has ruleId set 
                if (lib.tokenMaxTradingVolumeStorage().tokenMaxTradingVolume[action]) {
                    _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
                /// else if conditional used for tokenMaxTrading as there is only one ruleId used for this rule 
                } else if (lib.tokenMaxTradingVolumeStorage().tokenMaxTradingVolume[ActionTypes.SELL]) {
                    _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
                }
                /// tokenMinTxSize SELL Side 
                if (lib.tokenMinTxSizeStorage().tokenMinTxSize[ActionTypes.SELL].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
                /// tokenMaxDailyTrades SELL Side
                if (lib.tokenMaxDailyTradesStorage().tokenMaxDailyTrades[ActionTypes.SELL].active) _checkTokenMaxDailyTradesRule(action, _tokenId);
            } else { /// custodial buy 
                if (lib.tokenMaxTradingVolumeStorage().tokenMaxTradingVolume[action]) {
                    _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
                }
            }
            /// tokenMinTxSize BUY
            if (lib.tokenMinTxSizeStorage().tokenMinTxSize[action].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
            /// tokenMaxDailyTrades BUY
            if (lib.tokenMaxDailyTradesStorage().tokenMaxDailyTrades[action].active) _checkTokenMaxDailyTradesRule(action, _tokenId);```
            
Should be:


```
            if (_from != _sender){ /// non custodial buy 
                /// else if conditional used for tokenMaxTrading as there is only one ruleId used for this rule 
                if (lib.tokenMaxTradingVolumeStorage().tokenMaxTradingVolume[ActionTypes.SELL]) {
                    _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
                }
                /// tokenMinTxSize SELL Side 
                if (lib.tokenMinTxSizeStorage().tokenMinTxSize[ActionTypes.SELL].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
                /// tokenMaxDailyTrades SELL Side
                if (lib.tokenMaxDailyTradesStorage().tokenMaxDailyTrades[ActionTypes.SELL].active) _checkTokenMaxDailyTradesRule(action, _tokenId);
            }
            if (lib.tokenMaxTradingVolumeStorage().tokenMaxTradingVolume[action]) _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
            /// tokenMinTxSize BUY
            if (lib.tokenMinTxSizeStorage().tokenMinTxSize[action].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
            /// tokenMaxDailyTrades BUY
            if (lib.tokenMaxDailyTradesStorage().tokenMaxDailyTrades[action].active) _checkTokenMaxDailyTradesRule(action, _tokenId);
 ```
 
 I find this much easier to read and its probably a little more efficient.

File: src/client/token/handler/diamond/ERC20NonTaggedRuleFacet.sol [hunk 3]
```diff
@@ -23,13 +23,67 @@ contract ERC20NonTaggedRuleFacet is AppAdministratorOrOwnerOnlyDiamondVersion, H
...
+            /// tokenMinTxSize BUY
+            if (lib.tokenMinTxSizeStorage().tokenMinTxSize[ActionTypes.BUY].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
+            } else { /// custodial sell 
+                /// tokenMaxTradingVolume SELL 
+                if (lib.tokenMaxTradingVolumeStorage().tokenMaxTradingVolume[action]) {
+                    _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
+                }
+            }
+            /// tokenMinTxSize SELL
+            if (lib.tokenMinTxSizeStorage().tokenMinTxSize[action].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
```
Comment (VoR0220) [inline src/client/token/handler/diamond/ERC20NonTaggedRuleFacet.sol:84]: Similar here as above.

File: src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol (hunk 1 above)
Comment (VoR0220) [reply src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol:86]: Cracked it, see above.

File: src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol [hunk 4]
```diff
@@ -41,23 +41,84 @@ contract ERC721NonTaggedRuleFacet is
...
+            } else { /// custodial sell 
+                /// tokenMaxTradingVolume SELL 
+                if (lib.tokenMaxTradingVolumeStorage().tokenMaxTradingVolume[action]) {
+                    _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
+                }
+            }
+            /// tokenMinTxSize SELL
+            if (lib.tokenMinTxSizeStorage().tokenMinTxSize[action].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
+            /// tokenMaxDailyTrades SELL
+            if (lib.tokenMaxDailyTradesStorage().tokenMaxDailyTrades[action].active) _checkTokenMaxDailyTradesRule(action, _tokenId);
```
Comment (VoR0220) [inline src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol:115]: here too

**Review by oscarsernarosero — APPROVED** [review]
Looks great. I seconded RJ comment and left a solution proposal.

File: src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol [hunk 5]
```diff
@@ -41,23 +41,84 @@ contract ERC721NonTaggedRuleFacet is
...
+            /// tokenMinTxSize BUY Side 
+            if (lib.tokenMinTxSizeStorage().tokenMinTxSize[ActionTypes.BUY].active) _checkTokenMinTxSizeRule(_amount, action, handlerBase);
+            /// tokenMaxDailyTrades BUY Side
+            if (lib.tokenMaxDailyTradesStorage().tokenMaxDailyTrades[ActionTypes.BUY].active) _checkTokenMaxDailyTradesRule(action, _tokenId);
+            } else { /// custodial sell 
+                /// tokenMaxTradingVolume SELL 
+                if (lib.tokenMaxTradingVolumeStorage().tokenMaxTradingVolume[action]) {
+                    _checkTokenMaxTradingVolumeRule(_amount, handlerBase);
+                }
+            }
```
Comment (VoR0220) [inline src/client/token/handler/diamond/ERC721NonTaggedRuleFacet.sol:111]: Basically the rule should be, if you can determine there is a non custodial buy/sell, then do the opposite action and apply those rules if they exist, and then fallthrough to carry on normally if it isn't a non custodial and carry on business as normal.

**Review by VoR0220 — APPROVED** [review]
After a long discussion and trial and error session, we're taking the L on the hard to parse code. Something to fix in the future.
